Support Row Type in New Parquet Reader

zhenxiao opened the pull request.

zhenxiao commented:
@dain would you please review when you are free?

dain commented:
@nezihyigitbasi can you review this also?

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetTypeUtils.java:
```diff
@@ -17,14 +17,33 @@
+        for (String[] path : paths) {
+            PrimitiveType primitiveType = schema.getType(path).asPrimitiveType();
+            columns.add(new RichColumnDescriptor(
+                    path,
```
nit: move `path,` to previous line

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java:
```diff
@@ -200,7 +200,6 @@ private void readNextPage()
     {
         page = pageReader.readPage();
         validateParquet(page != null, "Not enough values to read in column chunk");
-        pageValueCount = page.getValueCount();
```
Do we need the `pageValueCount` field at all? I guess in all places that we read this field we have either access to the page itself or we pass the count as an argument, so seems like we can just remove it.

zhenxiao replied:
removed

nezihyigitbasi commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java:
```diff
@@ -279,7 +279,7 @@ public void testParquetPageSource(int rowCount)
         List<TestColumn> testColumns = getTestColumnsSupportedByParquet();
         testColumns = testColumns.stream()
-                .filter(column -> column.getObjectInspector().getCategory() == Category.PRIMITIVE)
+                .filter(column -> column.getObjectInspector().getCategory() == Category.PRIMITIVE || column.getName().equals("t_struct_bigint") || column.getName().equals("t_struct_null") || column.getName().equals("t_struct_non_nulls_after_nulls") || column.getName().equals("t_nested_struct_non_nulls_after_nulls"))
```
this is a long predicate, how about creating a list with the new test column names and then do a contains check?

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -213,6 +214,59 @@ public Page getNextPage()
+                }
+                pathFields.add(pathField);
+                int index = fieldNames.indexOf(pathField);
+                if (index != -1) {
```
is it even possible for `index` to be `-1`? If not, we can just assert that` index != -1`

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -213,6 +214,59 @@ public Page getNextPage()
+                    else {
+                        RichColumnDescriptor columnDescriptor = getColumns(requestedSchema).get(i);
+                        try {
+                            blocks[index] = parquetReader.readBlock(columnDescriptor, fieldType);
```
why don't we load this block lazily like the other blocks?

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -213,6 +214,59 @@ public Page getNextPage()
+        for (int i = 0; i < batchSize; i++) {
+            offsets[i] = (i + 1) * parameters.size();
+        }
+        return new ArrayBlock(interleavedBlock, Slices.wrappedIntArray(offsets), 0, Slices.wrappedBooleanArray(nullVector));
```
static import

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -213,6 +214,59 @@ public Page getNextPage()
+    {
+        Block[] blocks = new Block[parameters.size()];
+        List<String> fieldNames = new ArrayList<>();
+        boolean[] nullVector = new boolean[batchSize];
```
`nullVector` isn't filled in at all

nezihyigitbasi replied:
@zhenxiao this is still not filled in.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -18,28 +18,35 @@
 import com.facebook.presto.spi.ConnectorPageSource;
```
how about a commit message like "Add support for row type in the new Parquet reader"?

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -169,33 +178,25 @@ public Page getNextPage()
-                        fieldIndex = fileSchema.getFieldIndex(columnNames.get(fieldId));
+                    Type type = types.get(fieldId);
+                    if (ROW.equals(type.getTypeSignature().getBase())) {
+                        blocks[fieldId] = buildRowBlock(type.getTypeSignature().getParameters(), batchSize, 1);
```
the constant `1` is confusing, you can create a recursive helper method to pass that to `buildRowBlock`.

nezihyigitbasi commented:
Don't know whether the root cause is the same, but with a table that has a single `row` column with a `map` in it, when I select the `map` field I get the same exception with the second stacktrace above (`HIVE_CURSOR_ERROR` ...).

You can create the table with

```
create table struct_parquet_tests_2 
as select CAST(row(map(array[1,2,3,4,5],array['a','b','c','d','e'])) 
               AS row(col0 map<integer, varchar>)) x;
```

zhenxiao commented:
thank you @nezihyigitbasi 
The complex types bug is because of it having struct of arrays or struct of maps, while array and maps are not supported yet. I just added checks to fall back to the old reader when there are nested arrays or maps
For column rename, do not have a nice testcase or solution, I updated it, could you please help me test it?
other comments addressed

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -240,4 +244,18 @@ private static boolean columnTypeSupported(List<HiveColumnHandle> columns)
         return regularColumnsSupported && partitionColumnsSupported;
     }
+
+    private static boolean supportType(TypeSignature typeSignature)
```
we can rename this method as `isTypeSupported`

nezihyigitbasi commented:
Column rename still fails with a NPE. This fixes it:

```
diff --git a/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java b/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java
index f9387be..d01fe60 100644
--- a/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java
+++ b/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java
@@ -195,7 +195,7 @@ public class ParquetPageSource
                         else {
                             fieldIndex = hiveColumnIndexes[fieldId];
                         }
-                        blocks[fieldId] = new LazyBlock(batchSize, new ParquetBlockLoader(getDescriptor(fileSchema, requestedSchema, fileSchema.getPaths().get(fieldIndex)[0]), type));
+                        blocks[fieldId] = new LazyBlock(batchSize, new ParquetBlockLoader(getDescriptor(fileSchema, requestedSchema, fileSchema.getFields().get(fieldIndex).getName()), type));
                     }
                 }

```

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -213,6 +214,59 @@ public Page getNextPage()
+                    else {
+                        RichColumnDescriptor columnDescriptor = getDescriptor(fileSchema, requestedSchema, pathField);
+                        try {
+                            blocks[index] = parquetReader.readBlock(columnDescriptor, fieldType);
```
why don't you return a `LazyBlock` here?

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -240,4 +244,18 @@ private static boolean columnTypeSupported(List<HiveColumnHandle> columns)
+        else if (ROW.equals(typeSignature.getBase())) {
+            return typeSignature.getParameters().stream()
+                .map(TypeSignatureParameter::getTypeSignatureOrNamedTypeSignature)
+                .map(Optional::get)
```
check for presence before calling `get` on an `Optional`

```
.filter(Optional::isPresent)
.map(Optional::get)
```

zhenxiao commented:
@nezihyigitbasi thanks a lot, get comments addressed

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -213,6 +214,58 @@ public Page getNextPage()
+        for (int i = 0; i < batchSize; i++) {
+            offsets[i] = (i + 1) * parameters.size();
+        }
+        return new ArrayBlock(interleavedBlock, wrappedIntArray(offsets), 0, allocate(batchSize));
```
nullVector is treated as {fieldA=null, fieldB=null...}
unlike ORC, there is not presentBit, not so easy to set value for nullVector for lazyBlock
@nezihyigitbasi @dain what do you think?

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java:
```diff
@@ -213,6 +216,53 @@ public Page getNextPage()
+            }
+            else {
+                try {
+                    blocks[i] = parquetReader.readBlock(columnDescriptor, fieldType);
```
No LazyBlock for Struct's sub-blocks, only LazyBlock for first level columns
Only scanning necessary sub-blocks for requested nest struct is done via nested column pruning: 
https://github.com/prestodb/presto/pull/5547

zhenxiao commented:
@nezihyigitbasi @dain get comments addressed, tested and running OK now, would you please review when you are free?

zhenxiao commented:
added schema evolution support
refactor code for supporting MAP and ARRAY in following steps
@nezihyigitbasi @dain would you please review when you are free?

dain commented on presto-hive/src/test/java/com/facebook/presto/hive/parquet/ParquetTester.java:
```diff
@@ -172,16 +174,19 @@ private static void assertFileContents(JobConf jobConf,
                                                         parquetMetadata.getBlocks(),
-                                                        dataSource);
+                                                        dataSource,
+                                                        typeManager);
```
this is formatted wrong

dain commented:
@nezihyigitbasi can you review the latest changes and if they look good to you, click the approve button (or at least remove the changes requested flag)

zhenxiao commented:
thank you @dain @nezihyigitbasi 
I get comments addressed and rebased on the current master, could you please take another look when you are free?

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -240,4 +245,19 @@ private static boolean columnTypeSupported(List<HiveColumnHandle> columns)
+                .map(Optional::get)
+                .allMatch(ParquetPageSourceFactory::isTypeSupported);
+        }
+        return true;
```
It looks like the code supports all primitive types now, but say it didn't support `DATE`, then this code would allow that though.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetTypeUtils.java:
```diff
@@ -14,17 +14,72 @@
+                }
+            }
+        }
+        return null;
```
I suggest returning `Optional<RichColumnDescriptor>` here as I see not all the callers handle the case when this returns null. e.g., `ParquetPageSource` passes the return of this call to `ParquetBlockLoader`, which doesn't check for nulls and pass it to `parquetReader.readPrimitive`. Making this optional should catch all these places.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java:
```diff
@@ -107,7 +126,42 @@ private boolean advanceToNextRowGroup()
+    {
+        List<TypeSignatureParameter> parameters = type.getTypeSignature().getParameters();
+        Block[] blocks = new Block[parameters.size()];
+        List<String> fieldNames = new ArrayList<>();
```
`fieldNames` is unused.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetTypeUtils.java:
```diff
@@ -14,17 +14,72 @@
+        try {
+            return fileSchema.getFieldIndex(name);
+        }
+        catch (InvalidRecordException e) {
```
when you get an `InvalidRecordException` it means that `name` doesn't exist in the schema, so isn't the below for loop guaranteed to find nothing on exception? Any case that for loop handles?

zhenxiao replied:
it is handling for ignoreCase name lookup.
Presto takes metastore field names in all lowercase, parquet having both lowercase and upper case. This loop try to find ignore case if could not find wth case

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java:
```diff
@@ -107,7 +126,42 @@ private boolean advanceToNextRowGroup()
     }
 
-    public Block readBlock(ColumnDescriptor columnDescriptor, Type type)
+    public Block readStruct(Type type, List<String> path)
```
`readStruct` modifies the `path` list passed to it. Makes it hard to track the state when reading the code, so why don't we make a copy here?

zhenxiao commented:
thank you @nezihyigitbasi 
get comments addressed

nezihyigitbasi commented:
@dain looks good. Thanks @zhenxiao!

dain commented:
Merged, thanks!
